### DISCUSSION TITLE
Perofrmans test error

### DISCUSSION THREAD

**Discussion by AhmetNuri** [HashLoad/horse#287] [General]
hello, I did a performance test on a new horse server that sent 100,000 requests in a row and measures the time of the incoming replies, but it gave an error after 17000 records.
this test
I also did it on indy rest server, MARS Rest server, Datasnap RestServer, mORMot Rest Server. All other vehicles successfully completed the test. Is there any additional setting you think I need to do?

**Comment by viniciussanchez** [comment]
What kind of application did you create? Console? Could you put the code here?

**Reply by AhmetNuri** [reply]
var
   i : integer;
  sw: TStopwatch;
  oldWorking, kacak: Integer;
  Satir: TStringList;
  ResultStr: string;
begin
  Memo1.Lines.Add('Start Test         :' ;
  ResultStr := '';
  kacak := 0;
  Satir := TStringList.Create;
  sw := TStopwatch.Create;
  sw.Reset;
  sw.Start;
  oldWorking := -1;
//  RESTClient1.BaseURL := 'http://127.0.0.1:8080/rest/default/HelloWorld';
  Memo1.Lines.Add('Hourse Indy Test');
  try
  for I := 1 to 100000 do
  begin
    if ResultStr = '' then
    begin
      try
        ResultStr := IndyRequest(HourseURL);
      finally

        Satir.Add(ResultStr);
        ResultStr := '';
        Application.ProcessMessages;
      end;

    end
    else
    begin
      Application.ProcessMessages;
      kacak := kacak + 1;
    end;

  end;
    except on E: Exception do
   Memo1.Lines.Add(e.Message + '  ' + i.ToString)
  end;
  sw.Stop;
  Memo1.Lines.Add('Respone Count  :' + Satir.Count.ToString);
  Memo1.Lines.Add('Wait  Count         :' + kacak.ToString);
  Memo1.Lines.Add('Time(ms)         :' + sw.ElapsedMilliseconds.ToString);
  Memo1.Lines.Add('END        :' + TimeToStr(Now));

**Comment by acarlomagno** [comment]
Hello,
these days I am also testing the performance of Horse ... I use Apache Benchmark tool to do the tests.
https://httpd.apache.org/docs/2.4/programs/ab.html

My server is an Ubuntu 20.04 (4vCPU and 8gb Ram) with Docker + NGINX + SSL (Let's Encrypt) and Horse in reverse proxy.

I hid my data with xxxxxxxxxxx

I tried from another server to call Horse like this:
ab -c1000 -n100000 https: // xxxxxxxxxxxxx / ping 
and the server return a simple json response 
requesting 100,000 requests with 1000 simultaneous calls.

this is the result: ... there are 205 failed but it was a nginx ssl issue

Server Software:        nginx/1.21.6
Server Hostname:        xxxxxxxxxxxxxxxxx
Server Port:            443
SSL/TLS Protocol:       TLSv1.2,ECDHE-RSA-AES256-GCM-SHA384,4096,256
TLS Server Name:        xxxxxxxxxxxxxxxxx

Document Path:          /ping
Document Length:        18 bytes

Concurrency Level:      1000
Time taken for tests:   252.530 seconds
Complete requests:      100000
Failed requests:        205
   (Connect: 0, Receive: 0, Length: 205, Exceptions: 0)
Total transferred:      22653465 bytes
HTML transferred:       1796310 bytes
Requests per second:    395.99 [#/sec] (mean)
Time per request:       2525.301 [ms] (mean)
Time per request:       2.525 [ms] (mean, across all concurrent requests)
Transfer rate:          87.60 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0 2066 2217.6   1407   95532
Processing:    23  420 8133.6     49  188428
Waiting:        0   50  34.5     48    1184
Total:        427 2486 8339.7   1461  188428

Percentage of the requests served within a certain time (ms)
  50%   1461
  66%   1671
  75%   2352
  80%   2447
  90%   3278
  95%   4688
  98%   8573
  99%   9559
 100%  188428 (longest request)

**Reply by AhmetNuri** [reply]
Apache has too many external parameters.
which makes it difficult to compare between libraries.
I prepared my own test environment for it. Everything is the same except the libraries used are different
